geowave-analytics/src/main/java/mil/nga/giat/geowave/analytics/kmeans/mapreduce/KMeansMapReduce.java

```diff
@@ -0,0 +1,296 @@
+			catch (final Exception e1) {
+				throw new IOException(
+						e1);
+			}
+
+			try {
+				centroidExtractor = config.getInstance(
+						CentroidParameters.Centroid.EXTRACTOR_CLASS,
+						KMeansMapReduce.class,
+						CentroidExtractor.class,
+						SimpleFeatureCentroidExtractor.class);
+			}
+			catch (final Exception e1) {
+				throw new IOException(
+						e1);
+			}
+
+			try {
+				itemWrapperFactory = config.getInstance(
+						CentroidParameters.Centroid.WRAPPER_FACTORY_CLASS,
+						KMeansMapReduce.class,
+						AnalyticItemWrapperFactory.class,
+						SimpleFeatureItemWrapperFactory.class);
+
+				itemWrapperFactory.initialize(config);
+			}
+			catch (final Exception e1) {
+				throw new IOException(
+						e1);
+			}
+		}
+	}
+
+	/** Optimization 
+	 */
+	public static class KMeansCombiner extends Reducer<GroupIDText, BytesWritable,GroupIDText, BytesWritable> {
+		private final GeoObjectDimensionValues geoObject = new GeoObjectDimensionValues();
+		private final BytesWritable outputValWritable = new BytesWritable();
+		
+		@Override
+		public void reduce(
+				final GroupIDText key,
+				final Iterable<BytesWritable> values,
+				final Reducer<GroupIDText, BytesWritable, GroupIDText, BytesWritable>.Context context ) throws IOException,
+				InterruptedException {
+			final GeoObjectDimensionValues totals = new GeoObjectDimensionValues();
+
+			for (final BytesWritable value : values) {
+				geoObject.fromBytes(value.getBytes());
+				totals.add(geoObject);
+			}
+			final byte[] outData = totals.toBytes();
+			outputValWritable.set(
+					outData,
+					0,
+					outData.length);
+			context.write(key,outputValWritable);
+		}
+	}
+	
+	public static class KMeansReduce extends
+			Reducer<GroupIDText, BytesWritable, GeoWaveOutputKey, Object>
+	{
+
+		protected CentroidManager<Object> centroidManager;
+		private final GeoObjectDimensionValues geoObject = new GeoObjectDimensionValues();
+
+		@Override
+		public void reduce(
+				final GroupIDText key,
+				final Iterable<BytesWritable> values,
+				final Reducer<GroupIDText, BytesWritable, GeoWaveOutputKey, Object>.Context context )
+				throws IOException,
+				InterruptedException {
+			final String centroidID = key.getID();
+			final String groupID = key.getGroupID();
+			final GeoObjectDimensionValues totals = new GeoObjectDimensionValues();
+
+			for (final BytesWritable value : values) {
+				geoObject.fromBytes(value.getBytes());
+				totals.add(geoObject);
+			}
+
+			final AnalyticItemWrapper<Object> centroid = getFeatureForCentroid(
+					centroidID,
+					groupID);
+
+			// do not update the cost, because this cost is associated with the
+			// centroid PRIOR to this update.
+			// centroid.setCost(totals.distance);
+			centroid.resetAssociatonCount();
+			centroid.incrementAssociationCount(totals.getCount());
+
+			final double ptCount = totals.getCount();
+			// mean
+			totals.x = totals.x / ptCount;
+			totals.y = totals.y / ptCount;
+			totals.z = totals.z / ptCount;
```

Comment:
some functions use z and others don't, just curious what methodology we use for determining whether to account for z?  is it a consistent approach?